File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_530/java/Solution.java
package com.hi.dhl.algorithms.leetcode._530.java;

import com.hi.dhl.algorithms.model.TreeNode;

import java.util.Stack;

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/10/12
 *     desc  :
 * </pre>
 */
class Solution {
    private int pre = Integer.MAX_VALUE;
    private int min = Integer.MAX_VALUE;

    public int getMinimumDifference(TreeNode root) {
        if (root == null) return 0;
        dfs(root);
        return min;
    }

    private void dfs(TreeNode root) {
        if (root == null) {
            return;
        }
        dfs(root.left);
        if (pre != Integer.MAX_VALUE) {
            min = Math.min(min, root.val - pre);
        }
        pre = root.val;
        dfs(root.right);
    }

    private void bfs(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        while (!stack.isEmpty() || root != null) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }

            TreeNode node = stack.pop();
            if (pre != Integer.MAX_VALUE) {
                min = Math.min(min, node.val - pre);
            }
            pre = node.val;
            root = node.right;
        }
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_530/kotlin/Solution.kt
package com.hi.dhl.algorithms.leetcode._530.kotlin

import com.hi.dhl.algorithms.model.TreeNode
import java.util.*

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/10/12
 *     desc  :
 * </pre>
 */

class Solution {
    var pre: Int = Int.MAX_VALUE
    var min: Int = Int.MAX_VALUE

    fun getMinimumDifference(root: TreeNode?): Int {
        if (root == null) return 0
        bfs(root)
        return min
    }

    fun dfs(root: TreeNode?) {
        if (root == null) {
            return
        }

        dfs(root.left)
        if (pre != Int.MAX_VALUE) {
            val sub = root.`val` - pre
            min = Math.min(min, sub)
        }
        pre = root.`val`
        dfs(root.right)
    }

    fun bfs(root: TreeNode?) {
        val stack = Stack<TreeNode>()
        var cur = root
        while (!stack.isEmpty() || cur != null) {
            while (cur != null) {
                stack.push(cur)
                cur = cur.left
            }

            val node = stack.pop()
            if (pre != Int.MAX_VALUE) {
                min = Math.min(min, node.`val` - pre)
            }
            pre = node.`val`
            cur = node.right
        }
    }

}